File: internal/domain/rpc/giveaways/giveaways.go
// Copyright (c) 2026 Adam Ousmer. All rights reserved.
// Proprietary. No license granted. See LICENSE.md.

// Package giveaways defines the private Transactions giveaway API. The
// structures in this package are deliberately independent of Ent and of the
// admin/dashboard implementations so every caller shares the same wire
// contract and no caller can smuggle provider references into a mutation.
package giveaways

import (
	"time"

	"ItsBagelBot/internal/domain/rpc"
)

// Private subjects are kept here so service wiring and dashboard callers do
// not hand-author privileged strings. The admin prefix is intentionally
// separate from the ordinary transactions RPC namespace.
const (
	AdminPrefix          = "bagel.rpc.admin.giveaways"
	MineSubject          = "bagel.rpc.transactions.giveaways.mine"
	UsersPoolSubject     = "bagel.rpc.internal.users.giveaway.pool"
	UsersCoverageSubject = "bagel.rpc.internal.users.giveaway.coverage"
	UsersPrepareSubject  = "bagel.rpc.internal.users.giveaway.prepare"
	UsersCommitSubject   = "bagel.rpc.internal.users.giveaway.commit"
	UsersCancelSubject   = "bagel.rpc.internal.users.giveaway.cancel"
	VerbCreate           = "create"
	VerbList             = "list"
	VerbPreview          = "preview"
	VerbFreeze           = "freeze"
	VerbDraw             = "draw"
	VerbGet              = "get"
	VerbRetry            = "retry"
	VerbAlerts           = "alerts"
	VerbHistory          = "history"
	VerbCapabilities     = "capabilities"
)

type Status string

const (
	StatusDraft     Status = "draft"
	StatusFrozen    Status = "frozen"
	StatusDrawn     Status = "drawn"
	StatusCancelled Status = "cancelled"
)

type AwardState string

const (
	AwardSelected    AwardState = "selected"
	AwardPreparing   AwardState = "preparing"
	AwardNeedsReview AwardState = "needs_review"
	AwardScheduled   AwardState = "scheduled"
	AwardActive      AwardState = "active"
	AwardCompleted   AwardState = "completed"
	AwardVoided      AwardState = "voided"
)

type BillingState string

const (
	BillingNotRequired BillingState = "not_required"
	BillingPending     BillingState = "pending"
	BillingProtected   BillingState = "protected"
	BillingUncertain   BillingState = "uncertain"
	BillingIncident    BillingState = "incident"
	BillingReconciled  BillingState = "reconciled"
)

type EmailState string

const (
	EmailQueued         EmailState = "queued"
	EmailAccepted       EmailState = "accepted"
	EmailFailed         EmailState = "failed"
	EmailUncertain      EmailState = "uncertain"
	EmailMissingContact EmailState = "missing_contact"
)

// Mutation identifies the authenticated operator and protects retries and
// stale dashboard tabs. ActorRole is informational; Transactions resolves the
// actor's role server-side and never trusts this field for authorization.
type Mutation struct {
	ActorID         string `json:"actor_id"`
	ActorRole       string `json:"actor_role,omitempty"`
	IdempotencyKey  string `json:"idempotency_key"`
	ExpectedVersion uint64 `json:"expected_version,omitempty"`
	Reason          string `json:"reason,omitempty"`
}

type Campaign struct {
	ID           string     `json:"id"`
	Title        string     `json:"title"`
	Reason       string     `json:"reason,omitempty"`
	RulesVersion string     `json:"rules_version"`
	WinnerCount  int        `json:"winner_count"`
	PrizeMonths  int        `json:"prize_months"`
	Status       Status     `json:"status"`
	CreatedBy    uint64     `json:"created_by"`
	Version      uint64     `json:"version"`
	CreatedAt    time.Time  `json:"created_at"`
	UpdatedAt    time.Time  `json:"updated_at"`
	FrozenAt     *time.Time `json:"frozen_at,omitempty"`
	DrawnAt      *time.Time `json:"drawn_at,omitempty"`
}

type Candidate struct {
	ID              string `json:"id,omitempty"`
	UserID          uint64 `json:"user_id"`
	Username        string `json:"username,omitempty"`
	Eligible        bool   `json:"eligible"`
	ExclusionReason string `json:"exclusion_reason,omitempty"`
	Eligibility     any    `json:"eligibility,omitempty"`
}

type PoolSummary struct {
	Total            int                 `json:"total"`
	Eligible         int                 `json:"eligible"`
	Excluded         int                 `json:"excluded"`
	Free             int                 `json:"free"`
	Premium          int                 `json:"premium"`
	Subscribers      int                 `json:"subscribers"`
	RequestedWinners int                 `json:"requested_winners"`
	PrizeMonths      int                 `json:"prize_months"`
	TotalPrizeMonths string              `json:"total_prize_months"`
	Exclusions       PoolExclusionCounts `json:"exclusions"`
}

type PoolExclusionCounts struct {
	Banned       int `json:"banned"`
	Inactive     int `json:"inactive"`
	NotOnboarded int `json:"not_onboarded"`
	TestAccount  int `json:"test_account"`
	VIP          int `json:"vip"`
	CurrentStaff int `json:"current_staff"`
}

type CreateRequest struct {
	Mutation
	Title       string `json:"title"`
	Reason      string `json:"reason,omitempty"`
	WinnerCount int    `json:"winner_count"`
	PrizeMonths int    `json:"prize_months"`
}

type CreateReply struct {
	Campaign     *Campaign    `json:"campaign,omitempty"`
	Capabilities Capabilities `json:"capabilities"`
	rpc.Refusal
}

type PreviewRequest struct {
	Mutation
	CampaignID  string `json:"campaign_id,omitempty"`
	WinnerCount int    `json:"winner_count,omitempty"`
	PrizeMonths int    `json:"prize_months,omitempty"`
}
type PreviewReply struct {
	Campaign     *Campaign    `json:"campaign,omitempty"`
	Candidates   []Candidate  `json:"candidates,omitempty"`
	Summary      PoolSummary  `json:"summary"`
	PoolDigest   string       `json:"pool_digest,omitempty"`
	Capabilities Capabilities `json:"capabilities"`
	rpc.Refusal
}

type FreezeRequest struct {
	Mutation
	CampaignID string `json:"campaign_id"`
	PoolDigest string `json:"pool_digest"`
}
type FreezeReply struct {
	Campaign     *Campaign    `json:"campaign,omitempty"`
	Summary      PoolSummary  `json:"summary"`
	Capabilities Capabilities `json:"capabilities"`
	rpc.Refusal
}

type DrawRequest struct {
	Mutation
	CampaignID string `json:"campaign_id"`
	PoolDigest string `json:"pool_digest,omitempty"`
}
type DrawReply struct {
	Draw         *Draw        `json:"draw,omitempty"`
	Awards       []Award      `json:"awards,omitempty"`
	Capabilities Capabilities `json:"capabilities"`
	rpc.Refusal
}

type Draw struct {
	ID               string    `json:"id"`
	CampaignID       string    `json:"campaign_id"`
	OperationKey     string    `json:"operation_key"`
	PoolDigest       string    `json:"pool_digest"`
	AlgorithmVersion string    `json:"algorithm_version"`
	WinnerIDs        []uint64  `json:"winner_ids"`
	ActorID          uint64    `json:"actor_id"`
	CreatedAt        time.Time `json:"created_at"`
}

type Award struct {
	ID                 string       `json:"id"`
	CampaignID         string       `json:"campaign_id"`
	DrawID             string       `json:"draw_id"`
	UserID             uint64       `json:"user_id"`
	Ordinal            uint64       `json:"ordinal"`
	PrizeMonths        int          `json:"prize_months"`
	IntervalRule       string       `json:"interval_rule"`
	State              AwardState   `json:"state"`
	BillingState       BillingState `json:"billing_state"`
	EmailState         EmailState   `json:"email_state"`
	PlannedStart       *time.Time   `json:"planned_start,omitempty"`
	PlannedEnd         *time.Time   `json:"planned_end,omitempty"`
	ConfirmedStart     *time.Time   `json:"confirmed_start,omitempty"`
	ConfirmedEnd       *time.Time   `json:"confirmed_end,omitempty"`
	GrantID            string       `json:"grant_id,omitempty"`
	BillingOperationID string       `json:"billing_operation_id,omitempty"`
	FailureReason      string       `json:"failure_reason,omitempty"`
	RetryCount         uint64       `json:"retry_count"`
	Version            uint64       `json:"version"`
	SelectedAt         time.Time    `json:"selected_at"`
	UpdatedAt          time.Time    `json:"updated_at"`
}

type AwardLookupRequest struct {
	Mutation
	AwardID string `json:"award_id"`
}
type AwardLookupReply struct {
	Award *Award `json:"award,omitempty"`
	rpc.Refusal
}
type AwardRetryRequest struct {
	Mutation
	AwardID string `json:"award_id"`
}
type AwardRetryReply struct {
	Award        *Award       `json:"award,omitempty"`
	Capabilities Capabilities `json:"capabilities"`
	rpc.Refusal
}

type ListRequest struct {
	Mutation
	Status Status `json:"status,omitempty"`
	Limit  int    `json:"limit,omitempty"`
}
type ListReply struct {
	Campaigns    []Campaign   `json:"campaigns,omitempty"`
	Capabilities Capabilities `json:"capabilities"`
	rpc.Refusal
}

type GetRequest struct {
	Mutation
	CampaignID string `json:"campaign_id"`
}
type GetReply struct {
	Campaign     *Campaign    `json:"campaign,omitempty"`
	Awards       []Award      `json:"awards,omitempty"`
	Candidates   []Candidate  `json:"candidates,omitempty"`
	Draw         *Draw        `json:"draw,omitempty"`
	Capabilities Capabilities `json:"capabilities"`
	rpc.Refusal
}

type MineRequest struct {
	UserID string `json:"user_id"`
}

func (r MineRequest) Requested() string { return r.UserID }

type MineReply struct {
	Awards []Award `json:"awards,omitempty"`
	rpc.Refusal
}

type HistoryRequest struct {
	Mutation
	UserID string `json:"user_id"`
	Limit  int    `json:"limit,omitempty"`
}
type HistoryReply struct {
	Campaigns    []Campaign   `json:"campaigns,omitempty"`
	Awards       []Award      `json:"awards,omitempty"`
	Capabilities Capabilities `json:"capabilities"`
	rpc.Refusal
}

type ListAlertsRequest struct {
	Mutation
	UnresolvedOnly bool `json:"unresolved_only"`
	Limit          int  `json:"limit,omitempty"`
}
type Alert struct {
	ID               string     `json:"id"`
	AwardID          string     `json:"award_id"`
	OperationID      string     `json:"operation_id,omitempty"`
	Category         string     `json:"category"`
	State            string     `json:"state"`
	Message          string     `json:"message"`
	AffectedBoundary *time.Time `json:"affected_boundary,omitempty"`
	FirstSeenAt      time.Time  `json:"first_seen_at"`
	LastSeenAt       time.Time  `json:"last_seen_at"`
	AcknowledgedBy   uint64     `json:"acknowledged_by,omitempty"`
	AcknowledgedAt   *time.Time `json:"acknowledged_at,omitempty"`
	ResolvedAt       *time.Time `json:"resolved_at,omitempty"`
}
type ListAlertsReply struct {
	Alerts       []Alert      `json:"alerts,omitempty"`
	Capabilities Capabilities `json:"capabilities"`
	rpc.Refusal
}

type Capabilities struct {
	NewAwardsEnabled     bool   `json:"new_awards_enabled"`
	SchedulingEnabled    bool   `json:"scheduling_enabled"`
	ProviderMutations    bool   `json:"provider_mutations_enabled"`
	IntervalRuleVerified bool   `json:"interval_rule_verified"`
	Reason               string `json:"reason,omitempty"`
}

type CapabilitiesReply struct {
	Capabilities Capabilities `json:"capabilities"`
	rpc.Refusal
}

type CapabilitiesRequest struct{ Mutation }

// ProviderMutationGate makes the launch restriction explicit. A false gate
// permits read/verification calls but refuses pause/resume mutations.
type ProviderMutationGate struct {
	Enabled bool   `json:"enabled"`
	Reason  string `json:"reason,omitempty"`
}

File: web/admin/src/lib/server/giveaways.ts
import { rpc } from '@bagel/kit/server/nats';
import type { GiveawayEligibility, GiveawaySummary } from '@bagel/kit/giveaway';
import { randomUUID } from 'node:crypto';

const READ_TIMEOUT = 5000;

export type GiveawayStatus = 'draft' | 'frozen' | 'drawn' | 'cancelled' | 'complete';
export type AwardState = 'selected' | 'preparing' | 'needs_review' | 'scheduled' | 'active' | 'completed' | 'voided';
export type BillingState = 'not_required' | 'pending' | 'protected' | 'uncertain' | 'incident' | 'reconciled';
export type EmailState = 'queued' | 'accepted' | 'missing_contact' | 'failed' | 'uncertain';

export type GiveawayCapabilities = {
  newAwardsEnabled: boolean;
  schedulingEnabled: boolean;
  providerMutationsEnabled: boolean;
  intervalRuleVerified: boolean;
  reason?: string;
};

export type GiveawayWire = {
  id: string; title: string; reason: string; status: GiveawayStatus; winnerCount: number; prizeMonths: number;
  summary?: GiveawaySummary; eligible?: GiveawayEligibility; createdAt: string; drawnAt?: string | null;
  pendingAwards?: number; version: number; capabilities?: GiveawayCapabilities;
};

export type GiveawayWinnerWire = {
  id: string; userId: string; login: string; displayName?: string; selectedAt: string; prizeMonths: number;
  awardState: AwardState; billingState: BillingState; startAt?: string | null; endAt?: string | null;
  paidThroughAt?: string | null; nextChargeAt?: string | null; providerVerifiedAt?: string | null;
  emailState: EmailState; emailWarning?: string | null; failureReason?: string | null; retryCount: number;
};

export type GiveawayAlertWire = {
  id: string; awardId: string; reason: string; pendingSince: string; upcomingChargeAt?: string | null; urgent?: boolean;
};

export type GiveawayDetailWire = GiveawayWire & {
  winners: GiveawayWinnerWire[]; eligible?: GiveawayEligibility; poolDigest?: string | null;
  eligibleCount?: number; selectionMethod?: 'random_draw'; algorithmVersion?: string | null; alerts?: GiveawayAlertWire[];
};

export type GiveawayPreviewWire = {
  eligible: GiveawayEligibility; exclusions: Record<string, number>; summary: GiveawaySummary;
  durationProtectionWarnings?: string[]; frozen?: boolean; poolDigest?: string; capabilities?: GiveawayCapabilities;
};

export type FreezeOperation = { actorId: string; campaignId: string; poolDigest: string; expectedVersion: number; idempotencyKey: string };
export type DrawOperation = { actorId: string; campaignId: string; expectedVersion: number; idempotencyKey: string };

type CampaignRaw = {
  id: string; title: string; reason?: string; rules_version: string; winner_count: number; prize_months: number;
  status: GiveawayStatus; created_by: number; version: number; created_at: string; updated_at: string;
  frozen_at?: string; drawn_at?: string;
};
type AwardRaw = {
  id: string; campaign_id: string; draw_id: string; user_id: number; ordinal: number; prize_months: number;
  interval_rule: string; state: AwardState; billing_state: BillingState; email_state: EmailState;
  planned_start?: string; planned_end?: string; confirmed_start?: string; confirmed_end?: string;
  grant_id?: string; billing_operation_id?: string; failure_reason?: string; retry_count: number;
  version: number; selected_at: string; updated_at: string;
};
type CandidateRaw = { id?: string; user_id: number; username?: string; eligible: boolean; exclusion_reason?: string; eligibility?: unknown };
type PoolExclusionCountsRaw = {
  banned: number; inactive: number; not_onboarded: number; test_account: number; vip: number; current_staff: number;
};
type PoolSummaryRaw = {
  total: number; eligible: number; excluded: number; free: number; premium: number; subscribers: number;
  requested_winners: number; prize_months: number; total_prize_months: string; exclusions: PoolExclusionCountsRaw;
};
type CapabilitiesRaw = {
  new_awards_enabled: boolean; scheduling_enabled: boolean; provider_mutations_enabled: boolean;
  interval_rule_verified: boolean; reason?: string;
};
type AlertRaw = {
  id: string; award_id: string; operation_id?: string; category: string; state: string; message: string;
  affected_boundary?: string; first_seen_at: string; last_seen_at: string; acknowledged_by?: number;
  acknowledged_at?: string; resolved_at?: string;
};
type CampaignReply = { campaign?: CampaignRaw; capabilities?: CapabilitiesRaw };
type ListReply = { campaigns?: CampaignRaw[]; capabilities?: CapabilitiesRaw };
type PreviewReply = { campaign?: CampaignRaw; candidates?: CandidateRaw[]; summary: PoolSummaryRaw; pool_digest?: string; capabilities: CapabilitiesRaw };
type DrawRaw = { pool_digest: string; algorithm_version: string };
type GetReply = { campaign?: CampaignRaw; awards?: AwardRaw[]; candidates?: CandidateRaw[]; draw?: DrawRaw; capabilities?: CapabilitiesRaw };
type DrawReply = { awards?: AwardRaw[]; draw?: DrawRaw; capabilities?: CapabilitiesRaw };
type MutationInput = { actorId: string; idempotencyKey?: string; expectedVersion?: number; reason?: string };

export function buildMutation(input: MutationInput): Record<string, string | number> {
  return {
    actor_id: input.actorId, idempotency_key: input.idempotencyKey ?? randomUUID(),
    ...(input.expectedVersion === undefined ? {} : { expected_version: input.expectedVersion }),
    ...(input.reason ? { reason: input.reason } : {})
  };
}

function capabilities(raw?: CapabilitiesRaw): GiveawayCapabilities | undefined {
  if (!raw) return undefined;
  return {
    newAwardsEnabled: raw.new_awards_enabled, schedulingEnabled: raw.scheduling_enabled,
    providerMutationsEnabled: raw.provider_mutations_enabled, intervalRuleVerified: raw.interval_rule_verified, reason: raw.reason
  };
}

export function mapCampaign(raw: CampaignRaw): GiveawayWire {
  return {
    id: raw.id, title: raw.title, reason: raw.reason ?? '', status: raw.status, winnerCount: raw.winner_count,
    prizeMonths: raw.prize_months, createdAt: raw.created_at, drawnAt: raw.drawn_at, version: raw.version
  };
}

function eligibility(summary: PoolSummaryRaw): GiveawayEligibility {
  return { eligible: summary.eligible, free: summary.free, premium: summary.premium, subscribers: summary.subscribers, excluded: summary.excluded };
}

function summary(raw: PoolSummaryRaw): GiveawaySummary {
  const numericTotal = Number(raw.total_prize_months);
  return { winners: raw.requested_winners, months: raw.prize_months, totalMonths: Number.isSafeInteger(numericTotal) ? numericTotal : raw.total_prize_months };
}

export function mapAward(raw: AwardRaw, login = String(raw.user_id)): GiveawayWinnerWire {
  return {
    id: raw.id, userId: String(raw.user_id), login, selectedAt: raw.selected_at,
    prizeMonths: raw.prize_months, awardState: raw.state, billingState: raw.billing_state,
    startAt: raw.confirmed_start ?? raw.planned_start, endAt: raw.confirmed_end ?? raw.planned_end,
    emailState: raw.email_state, failureReason: raw.failure_reason, retryCount: raw.retry_count
  };
}

const DRAW_ALGORITHM = 'crypto-rand-partial-fisher-yates-v1';

export function persistedEligibleCount(status: GiveawayStatus, candidates: ReadonlyArray<{ eligible: boolean }>): number | undefined {
  if (!['frozen', 'drawn', 'cancelled', 'complete'].includes(status) || candidates.length === 0) return undefined;
  return candidates.filter((candidate) => candidate.eligible).length;
}

export function selectionMethodForAlgorithm(algorithmVersion?: string): 'random_draw' | undefined {
  return algorithmVersion === DRAW_ALGORITHM ? 'random_draw' : undefined;
}

export function mapDrawMetadata(draw?: DrawRaw): { poolDigest?: string; algorithmVersion?: string; selectionMethod?: 'random_draw' } {
  if (!draw) return {};
  return {
    poolDigest: draw.pool_digest,
    algorithmVersion: draw.algorithm_version,
    selectionMethod: selectionMethodForAlgorithm(draw.algorithm_version)
  };
}

export function mapPreview(reply: PreviewReply): GiveawayPreviewWire {
  return {
    eligible: eligibility(reply.summary), exclusions: { ...reply.summary.exclusions }, summary: summary(reply.summary),
    poolDigest: reply.pool_digest, capabilities: capabilities(reply.capabilities)
  };
}

function mapAlert(raw: AlertRaw): GiveawayAlertWire {
  return { id: raw.id, awardId: raw.award_id, reason: raw.message, pendingSince: raw.first_seen_at, upcomingChargeAt: raw.affected_boundary };
}

function requireCampaign(campaign: CampaignRaw | undefined): CampaignRaw {
  if (!campaign) throw new Error('Giveaway service returned no campaign.');
  return campaign;
}

export async function giveawaysList(input: { actorId: string }): Promise<GiveawayWire[]> {
  const reply = await rpc<ListReply>('bagel.rpc.admin.giveaways.list', { ...buildMutation(input), limit: 100 }, READ_TIMEOUT);
  return (reply.campaigns ?? []).map(mapCampaign);
}

export async function giveawayGet(input: { actorId: string; campaignId: string }): Promise<GiveawayDetailWire> {
  const reply = await rpc<GetReply>('bagel.rpc.admin.giveaways.get', { ...buildMutation(input), campaign_id: input.campaignId }, READ_TIMEOUT);
  const campaign = mapCampaign(requireCampaign(reply.campaign));
  const logins = new Map((reply.candidates ?? []).map((candidate) => [String(candidate.user_id), candidate.username ?? String(candidate.user_id)]));
  const draw = mapDrawMetadata(reply.draw);
  return {
    ...campaign, winners: (reply.awards ?? []).map((award) => mapAward(award, logins.get(String(award.user_id)))),
    ...draw, eligibleCount: persistedEligibleCount(campaign.status, reply.candidates ?? []),
    capabilities: capabilities(reply.capabilities)
  };
}

export type GiveawayPreviewInput =
  | { actorId: string; campaignId: string }
  | { actorId: string; winnerCount: number; prizeMonths: number };

export async function giveawayPreview(input: GiveawayPreviewInput): Promise<GiveawayPreviewWire> {
  const request = {
    ...buildMutation(input),
    ...('campaignId' in input ? { campaign_id: input.campaignId } : { winner_count: input.winnerCount, prize_months: input.prizeMonths })
  };
  const reply = await rpc<PreviewReply>('bagel.rpc.admin.giveaways.preview', request, READ_TIMEOUT);
  return mapPreview(reply);
}

export async function giveawayFreeze(operation: FreezeOperation): Promise<void> {
  await rpc('bagel.rpc.admin.giveaways.freeze', { ...buildMutation(operation), campaign_id: operation.campaignId, pool_digest: operation.poolDigest }, READ_TIMEOUT);
}

export async function giveawayCreate(input: { actorId: string; title: string; reason: string; winnerCount: number; prizeMonths: number }): Promise<GiveawayWire> {
  const reply = await rpc<CampaignReply>('bagel.rpc.admin.giveaways.create', {
    ...buildMutation(input), title: input.title, reason: input.reason, winner_count: input.winnerCount, prize_months: input.prizeMonths,
  }, READ_TIMEOUT);
  return mapCampaign(requireCampaign(reply.campaign));
}

export async function giveawayDraw(operation: DrawOperation): Promise<{ awards: GiveawayWinnerWire[] }> {
  const reply = await rpc<DrawReply>('bagel.rpc.admin.giveaways.draw', { ...buildMutation(operation), campaign_id: operation.campaignId }, READ_TIMEOUT);
  return { awards: (reply.awards ?? []).map((award) => mapAward(award)) };
}

export async function giveawayRetry(input: { actorId: string; awardId: string }): Promise<void> {
  await rpc('bagel.rpc.admin.giveaways.retry', { ...buildMutation(input), award_id: input.awardId }, READ_TIMEOUT);
}

export async function giveawayAlerts(input: { actorId: string }): Promise<GiveawayAlertWire[]> {
  const reply = await rpc<{ alerts?: AlertRaw[] }>('bagel.rpc.admin.giveaways.alerts', { ...buildMutation(input), unresolved_only: true, limit: 100 }, READ_TIMEOUT);
  return (reply.alerts ?? []).map(mapAlert);
}

export async function giveawayHistory(input: { actorId: string; userId: string }): Promise<GiveawayWinnerWire[]> {
  const reply = await rpc<{ awards?: AwardRaw[] }>('bagel.rpc.admin.giveaways.history', { ...buildMutation(input), user_id: input.userId }, READ_TIMEOUT);
  return (reply.awards ?? []).map((award) => mapAward(award));
}
